libs/formats/src/zone/zone-validator.ts

```diff
@@ -0,0 +1,325 @@
+              `Volume ${volumeIndex} region ${regionIndex} polygon is self-intersecting`
+            );
+        }
+      }
+
+      if ((region.teleportPos == null) !== (region.teleportYaw == null))
+        throw new ZoneValidationError(
+          `Volume ${volumeIndex} region ${regionIndex} has a teleport pos or yaw specified but should have both or neither.`
+        );
+    }
+  }
+
+  function validateZone(
+    zone: Zone,
+    trackName: string,
+    zoneName: string,
+    volumes: Volume[],
+    requiresTele: boolean
+  ) {
+    if (!zone) throw new ZoneValidationError('Missing zones');
+
+    if (zone.volumeIndex < 0 || zone.volumeIndex >= volumes.length)
+      throw new ZoneValidationError(
+        `Track ${trackName} ${zoneName} has invalid volume index`
+      );
+
+    usedVolumes[zone.volumeIndex] = true;
+
+    if (requiresTele) {
+      const zoneRegions = volumes[zone.volumeIndex].regions;
+      for (const [regionIndex, region] of zoneRegions.entries()) {
+        if (region.teleportPos == null)
+          throw new ZoneValidationError(
+            `Track ${trackName} ${zoneName} must specify a teleport destination for each Volume region (missing region ${regionIndex})`
+          );
+        if (region.teleportYaw == null)
+          throw new ZoneValidationError(
+            `Track ${trackName} ${zoneName} must specify a teleport yaw for each Volume region (missing region ${regionIndex})`
+          );
+      }
+    }
+
+    totalZones++;
+  }
+
+  function validateTrack(track: Track, trackName: string, volumes: Volume[]) {
+    if (
+      !Array.isArray(track.zones.segments) ||
+      track.zones.segments.length === 0
+    )
+      throw new ZoneValidationError(`Track ${trackName} has no segments`);
+
+    if (track.zones.segments.length > MAX_TRACK_SEGMENTS)
+      throw new ZoneValidationError(`Track ${trackName} has too many segments`);
+
+    for (const [segmentIndex, segment] of track.zones.segments.entries()) {
+      if (!Array.isArray(segment?.checkpoints))
+        throw new ZoneValidationError(
+          `Track ${trackName} segment ${segmentIndex} has no checkpoints`
+        );
+
+      // We don't return the overall zone file currently this copy isn't
+      // strictly necessary, as but .shift modifies the underlying array,
+      // so safest to copy first just in case we care about final MapZones
+      // object in the future.
+      const checkpoints = structuredClone(segment.checkpoints);
+      const start = checkpoints.shift();
+
+      if (!start)
+        throw new ZoneValidationError(
+          `Track ${trackName} segment ${segmentIndex} has no start zone`
+        );
+
+      validateZone(
+        start,
+        trackName,
+        `segment ${segmentIndex} start zone`,
+        volumes,
+        true
+      );
+
+      // TODO: because of the .shift this doesn't include start zone. Make
+      // sure C++ version is the same.
+      if (checkpoints.length > MAX_SEGMENT_CHECKPOINTS)
```

Comment:
C++ does in fact count the first element in `checkpoints` toward the limit, which matters because the limit keeps `minorNum` exactly within the limits of a byte (when 1-indexing). C++ handles the segment start in the same loop now:
```c++
FOR_EACH_VEC( track.zones.segments, segmentIndex )
{
	const TimerSegmentDef &segment = track.zones.segments[segmentIndex];

	if ( segment.checkpoints.Count() > MAX_SEGMENT_CHECKPOINTS )
		return TimerDefsError( "Track %s segment %i has too many checkpoints", trackName, segmentIndex );

	FOR_EACH_VEC( segment.checkpoints, checkpointIndex )
	{
		if ( !CheckZone( segment.checkpoints[checkpointIndex], trackName,
			CFmtStr( "segment %i timer zone %i", segmentIndex, checkpointIndex ),
			volumes, volumeIsUsed, checkpointIndex == 0 ) )
			return false;
	}
}```

Reply:
Okay, check is now `checkpoints.length + 1 > MAX_SEGMENT_CHECKPOINTS`